power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/buffer.hpp

```diff
@@ -0,0 +1,61 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_CPP_BUFFER_HPP
+#define POWER_GRID_MODEL_CPP_BUFFER_HPP
+
+#include "basics.hpp"
+#include "handle.hpp"
+
+#include "power_grid_model_c/buffer.h"
+
+namespace power_grid_model_cpp {
+class Buffer {
+  public:
+    Buffer(MetaComponent const* component, Idx size)
+        : component_{component}, size_{size}, buffer_{handle_.call_with(PGM_create_buffer, component, size)} {};
+
+    RawDataPtr get() const { return buffer_.get(); }
+
+    static void set_nan(MetaComponent const* component, RawDataPtr buffer, Idx buffer_offset, Idx size) {
+        Handle handle{};
+        handle.call_with(PGM_buffer_set_nan, component, buffer, buffer_offset, size);
+    }
+    static void set_nan(Buffer& buffer, Idx buffer_offset, Idx size) {
+        buffer.handle_.call_with(PGM_buffer_set_nan, buffer.component_, buffer.buffer_.get(), buffer_offset, size);
+    }
+    void set_nan(Idx buffer_offset) { set_nan(*this, buffer_offset, size_); }
+
+    static void set_value(MetaAttribute const* attribute, Buffer& buffer, RawDataConstPtr src_ptr, Idx buffer_offset,
+                          Idx src_stride) {
+        buffer.handle_.call_with(PGM_buffer_set_value, attribute, buffer.buffer_.get(), src_ptr, buffer_offset,
+                                 buffer.size_, src_stride);
+    }
+    void set_value(MetaAttribute const* attribute, RawDataConstPtr src_ptr, Idx buffer_offset, Idx src_stride) {
+        set_value(attribute, *this, src_ptr, buffer_offset, src_stride);
+    }
+    static void get_value(MetaAttribute const* attribute, RawDataConstPtr data, RawDataPtr dest_ptr, Idx buffer_offset,
+                          Idx size, Idx dest_stride) {
+        Handle handle{};
+        handle.call_with(PGM_buffer_get_value, attribute, data, dest_ptr, buffer_offset, size, dest_stride);
+    }
+    static void get_value(MetaAttribute const* attribute, Buffer const& buffer, RawDataPtr dest_ptr, Idx buffer_offset,
+                          Idx size, Idx dest_stride) {
+        buffer.handle_.call_with(PGM_buffer_get_value, attribute, buffer.buffer_.get(), dest_ptr, buffer_offset, size,
+                                 dest_stride);
+    }
+    void get_value(MetaAttribute const* attribute, RawDataPtr dest_ptr, Idx buffer_offset, Idx dest_stride) const {
+        get_value(attribute, *this, dest_ptr, buffer_offset, size_, dest_stride);
+    }
+
+  private:
+    Handle handle_{};
+    MetaComponent const* component_;
+    Idx size_;
+    detail::UniquePtr<void, &PGM_destroy_buffer> buffer_;
```

Comment:
Now I start to wonder if there is a use-case to `Buffer` copyable. You can create new buffers and use `memcpy` to copy the data.

But let's see how the re-routing works and if this is needed we can always add.

Reply:
i would argue that that's an intentionality thing. I would say that you can then explicity call a `Buffer::copy()` function to achieve that. If we make `Buffer` copy-constructible or copy-assignable, then it becomes really easy for the user to make a typo and copy gigabytes of data. If needed, we can add that later but indeed, I don't think it's necessary as of now

Reply:
Okay fair point

Reply:
All the tests are now core-independent and at least for now copying a buffer was not needed. However, this can be easily added in the future.